src/components/icons/create-styled-icon.js

```diff
@@ -0,0 +1,91 @@
+import PropTypes from 'prop-types';
+import invariant from 'tiny-invariant';
+import styled from '@emotion/styled';
+import vars from '../../../materials/custom-properties';
+
+const iconSizes = {
+  small: 12,
+  medium: 16,
+  big: 24,
+};
+
+const getSizeStyle = size => {
+  switch (size) {
+    case 'scale':
+      return `
+        svg.scale:not(:root) {
+          width: 100%;
+          height: auto;
+        }
+      `;
+    case 'small':
+    case 'medium':
+    case 'big':
+      return `
+        width: ${iconSizes[size]}px;
+        height: ${iconSizes[size]}px;
+      `;
+    default: {
+      invariant(
+        size,
+        `ui-kit/Icon: the specified size '${size}' is not supported.`
+      );
+      return '';
+    }
+  }
+};
+
+const getColor = theme => {
+  if (!theme) return 'inherit';
+  switch (theme) {
+    case 'black':
+      return vars.colorBlack;
+    case 'grey':
+      return vars.colorGray60;
+    case 'white':
+      return vars.colorWhite;
+    case 'blue':
+      return vars.colorBlue;
+    case 'green':
+      return vars.colorGreen;
+    case 'green-light':
+      return vars.colorGreen40;
+    case 'orange':
+      return vars.colorOrange;
+    case 'red':
+      return vars.colorRed;
+    default: {
+      invariant(
+        theme,
+        `ui-kit/Icon: the specified theme '${theme}' is not supported.`
+      );
+      return 'inherit';
+    }
+  }
+};
+
+export default function createStyledIcon(Component, displayName) {
```

Comment:
Why do we need a factory function for this? We have quite a few of these in the app and they cause wasteful renders if I recall. Would it be an option to just have an `StyledIcon` component accepting an `iconComponent` prop and `iconDisplayName`?

Reply:
The function is only used to create the component. The exported component is already styled, there are wasteful rerenders